switches/http/http.cpp

```diff
@@ -0,0 +1,1280 @@
+                 << result_str << endl << "\tupdates " << pending->size() << " pending";
+            for (ResponseSet::iterator i = pending->begin(); i != pending->end(); i++)
+                cerr << " " << *i;
+            cerr << endl;
+        }
+        
+        for (ResponseSet::iterator i = pending->begin(); i != pending->end(); )
+        {
+            // i points to an HttpRequest* that is waiting for this variable value
+            if (verbose)
+                cerr << *i << " updating response var (" << variables->source << "," << variables->start << ")" << endl;
+            finishResponse(*i, 200, result_str);
+            pending->erase(i++);
+        }
+        if (verbose)
+        {
+            cerr << "\tcheck " << pendingVariables[address].size() << " pending";
+            for (ResponseSet::iterator i =
+                 pendingVariables[address].begin(); i != pendingVariables[address].end(); i++)
+                cerr << " " << *i;
+            cerr << endl;
+        }
+        sendAvailableResponses();
+    }
+    
+    // Incoming User Messages
+    void HttpInterface::incomingUserMsg(const UserMessage *userMsg)
+    {
+        if (verbose)
+            cerr << "incomingUserMsg msg ("<< userMsg->type <<","<< &userMsg->data <<")" << endl;
+        
+        // skip if event not known (yet, aesl probably not loaded)
+        try { commonDefinitions.events.at(userMsg->type); }
+        catch (const std::out_of_range& oor) { return; }
+        
+        if (commonDefinitions.events[userMsg->type].name.find(L"_update_vars ")==0)
+        {
+            // update variables
+        }
+        if (eventSubscriptions.size() > 0)
+        {
+            // set up SSE message
+            std::stringstream reply;
+            char this_event[128];
+            wstringtocstr(commonDefinitions.events[userMsg->type].name, this_event);
+            string event_name = std::string(this_event);
+            reply << "data: " << event_name;
+            for (size_t i = 0; i < userMsg->data.size(); ++i)
+                reply << " " << userMsg->data[i];
+            reply << "\r\n\r\n";
+            
+            // In the HTTP world we set up a stream of Server-Sent Events for this.
+            // Note that event name is in commonDefinitions.events[userMsg->type].name
+            
+            for (StreamEventSubscriptionMap::iterator subscriber = eventSubscriptions.begin();
+                 subscriber != eventSubscriptions.end(); ++subscriber)
+            {
+                if (subscriber->second.count("*") >= 1 || subscriber->second.count(event_name) >= 1)
+                {
+                    appendResponse(subscriber->first, 200, true, reply.str().c_str());
+                }
+            }
+        }
+    }
+
+    //-- Routing for HTTP requests ---------------------------------------------------------
+
+    void HttpInterface::routeRequest(HttpRequest* req)
+    {
```

Comment:
No, it goes on pendingResponses and will be freed in the future by the incoming Aseba message (with the variable's value) that provides the response. Otherwise it is freed by unscheduleResponse when the stream shuts down.